File: event-emitter/src/test/java/nl/haploid/octowight/service/EventChannelServiceTest.java
package nl.haploid.octowight.service;

import mockit.Injectable;
import mockit.Mocked;
import mockit.StrictExpectations;
import mockit.Tested;
import nl.haploid.octowight.AtomChangeEvent;
import nl.haploid.octowight.JsonMapper;
import nl.haploid.octowight.TestData;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.clients.producer.RecordMetadata;
import org.junit.Test;

import java.util.Arrays;
import java.util.List;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;

public class EventChannelServiceTest {

	@Tested
	private EventChannelService service;

	@Injectable
	private KafkaProducer<String, String> kafkaProducer;

	@Injectable
	private JsonMapper jsonMapper;

	@Test
	public void testSendEvents(final @Mocked Future<RecordMetadata> future1,
							   final @Mocked Future<RecordMetadata> future2) throws Exception {
		final AtomChangeEvent event1 = TestData.atomChangeEvent();
		final AtomChangeEvent event2 = TestData.atomChangeEvent();
		final List<AtomChangeEvent> events = Arrays.asList(event1, event2);
		final String message1 = TestData.message();
		final String message2 = TestData.message();
		new StrictExpectations() {{
			jsonMapper.serialize(event1);
			times = 1;
			result = message1;
			kafkaProducer.send((ProducerRecord<String, String>) any);
			times = 1;
			result = future1;
			jsonMapper.serialize(event2);
			times = 1;
			result = message2;
			kafkaProducer.send((ProducerRecord<String, String>) any);
			times = 1;
			result = future2;
			future1.get(anyLong, TimeUnit.SECONDS);
			times = 1;
			future2.get(anyLong, TimeUnit.SECONDS);
			times = 1;
		}};
		service.sendEvents(events);
	}
}

File: event-emitter/src/test/scala/nl/haploid/octowight/service/EventChannelServiceTest.scala
package nl.haploid.octowight.service

import java.util.concurrent.{Future, TimeUnit}

import nl.haploid.octowight._
import org.apache.kafka.clients.producer.{KafkaProducer, ProducerRecord, RecordMetadata}
import org.easymock.EasyMock
import org.springframework.test.util.ReflectionTestUtils

class EventChannelServiceTest extends AbstractTest {
  @Tested private[this] val eventChannelService: EventChannelService = null
  @Mocked private[this] val kafkaProducer: KafkaProducer[String, String] = null
  @Mocked private[this] val jsonMapper: JsonMapper = null

  override def beforeEach() = {
    super.beforeEach()
    ReflectionTestUtils.setField(eventChannelService, "topic", TestData.topic)
  }

  "Channel" should "send events" in {
    val future1 = mock[Future[RecordMetadata]]
    val future2 = mock[Future[RecordMetadata]]
    val event1 = TestData.atomChangeEvent
    val event2 = TestData.atomChangeEvent
    val events = List(event1, event2)

    expecting {
      val message1 = TestData.message
      val message2 = TestData.message
      val recordMetadata1 = new RecordMetadata(null, 1, 1)
      val recordMetadata2 = new RecordMetadata(null, 2, 2)
      jsonMapper.serialize(event1) andReturn message1 once()
      kafkaProducer.send(EasyMock.anyObject(classOf[ProducerRecord[String, String]])) andReturn future1 once()
      jsonMapper.serialize(event2) andReturn message2 once()
      kafkaProducer.send(EasyMock.anyObject(classOf[ProducerRecord[String, String]])) andReturn future2 once()
      future1.get(5, TimeUnit.SECONDS) andReturn recordMetadata1 once()
      future2.get(5, TimeUnit.SECONDS) andReturn recordMetadata2 once()
    }
    whenExecuting(jsonMapper, kafkaProducer, future1, future2) {
      eventChannelService.sendEvents(events)
    }
  }
}
